### PR TITLE
Improved kernel startup and failure telemetry

### PR DESCRIPTION
**Pull request by DonJayamanne**
This ensures we can better track failures per environment instead of checking if things fail in generate for a machine id.

* When sending the telemetry we check if the user actually executed a cell for a particular kernel (notebook)
* When user starts a cell, we re-send the telemetry that indicates whether the notebook was successfully started or not.

### CODE REVIEW

File: src/kernels/execution/cellExecution.ts [hunk 1]
```diff
@@ -301,9 +309,19 @@ export class CellExecution implements IDisposable {
         const props = { notebook: this.controller.notebookType === JupyterNotebookView };
         if (!CellExecution.sentExecuteCellTelemetry) {
             CellExecution.sentExecuteCellTelemetry = true;
-            sendTelemetryEvent(Telemetry.ExecuteCellPerceivedCold, this.stopWatchForTelemetry.elapsedTime, props);
+            sendKernelTelemetryEvent(
```
Comment (DonJayamanne) [inline src/kernels/execution/cellExecution.ts:312]: Track this per resource, so we have more information about the startup times, e.g. per kernal language, remote vs local, etc.

File: src/kernels/jupyter/launcher/notebookProvider.ts [hunk 2]
```diff
@@ -88,29 +85,13 @@ export class NotebookProvider implements INotebookProvider {
             await this.jupyterNotebookProvider.connect(serverOptions);
         }
         Cancellation.throwIfCanceled(options.token);
-        trackKernelResourceInformation(options.resource, {
-            kernelConnection: options.kernelConnection,
```
Comment (DonJayamanne) [inline src/kernels/jupyter/launcher/notebookProvider.ts:92]: Moved this into the Kernel.ts class as thats the top most layer where we start the kernels
Then its easier to look for these

File: src/kernels/jupyter/launcher/notebookProvider.ts [hunk 3]
```diff
@@ -88,29 +85,13 @@ export class NotebookProvider implements INotebookProvider {
...
+        return rawLocalKernel
             ? this.rawNotebookProvider!.createNotebook(
                   options.resource,
                   options.kernelConnection,
                   options.ui,
                   options.token
               )
             : this.jupyterNotebookProvider.createNotebook(options);
-
-        sendKernelTelemetryWhenDone(
```
Comment (DonJayamanne) [inline src/kernels/jupyter/launcher/notebookProvider.ts:104]: Moved this into the Kernel.ts class as thats the top most layer where we start the kernels
Then its easier to look for these

File: src/kernels/kernel.ts [hunk 4]
```diff
@@ -153,6 +153,18 @@ abstract class BaseKernel<TKernelExecution extends BaseKernelExecution> implemen
...
         this.disposables.push({ dispose: () => this._kernelSocket.unsubscribe() });
+        trackKernelResourceInformation(this.resourceUri, {
+            kernelConnection: this.kernelConnectionMetadata,
+            actionSource: this.creator,
+            disableUI: this.startupUI.disableUI
+        });
+        this.startupUI.onDidChangeDisableUI(() => {
+            if (!this.startupUI.disableUI) {
+                trackKernelResourceInformation(this.resourceUri, {
+                    disableUI: false
```
Comment (DonJayamanne) [inline src/kernels/kernel.ts:164]: Keep track of this flag for the resource

File: src/kernels/kernel.ts [hunk 5]
```diff
@@ -305,20 +317,35 @@ abstract class BaseKernel<TKernelExecution extends BaseKernelExecution> implemen
...
-                throw ex;
+            const stopWatch = new StopWatch();
+            trackKernelResourceInformation(this.resourceUri, {
+                kernelConnection: this.kernelConnectionMetadata,
+                actionSource: this.creator
             });
+
+            this._jupyterSessionPromise = this.createJupyterSession()
+                .then((session) => {
+                    sendKernelTelemetryEvent(
```
Comment (DonJayamanne) [inline src/kernels/kernel.ts:328]: Moved from `notebookProvider.ts`

File: src/kernels/kernel.ts [hunk 6]
```diff
@@ -349,11 +376,7 @@ abstract class BaseKernel<TKernelExecution extends BaseKernelExecution> implemen
             Cancellation.throwIfCanceled(this.startCancellation.token);
             await this.initializeAfterStart(session);
 
-            sendKernelTelemetryEvent(
```
Comment (DonJayamanne) [inline src/kernels/kernel.ts:352]: Moved this to the place where this method is called

File: src/kernels/telemetry/helper.ts [hunk 7]
```diff
@@ -68,7 +76,12 @@ export function trackKernelResourceInformation(resource: Resource, information:
     currentData.kernelLiveCount = information.kernelLiveCount || currentData.kernelLiveCount || 0;
     currentData.kernelInterpreterCount = information.kernelInterpreterCount || currentData.kernelInterpreterCount || 0;
     currentData.pythonEnvironmentCount = InterpreterCountTracker.totalNumberOfInterpreters;
-
+    if (information.userExecutedCell) {
+        currentData.userExecutedCell = true;
+    }
+    if (typeof information.disableUI === 'boolean') {
+        currentData.disableUI = information.disableUI;
```
Comment (DonJayamanne) [inline src/kernels/telemetry/helper.ts:83]: This way we now know whether the telemtry for a particular kernel is a result of user execution or not.
Previously it was for all kernels as we were only interested in whether things worked - at a user level.

**Review by IanMatthewHuff — APPROVED** [review]

**Pull request merged**